File: app/src/main/java/pl/karol202/paintplus/tool/selection/SelectionProperties.java
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package pl.karol202.paintplus.tool.selection;

import android.os.Bundle;
import android.view.*;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.Button;
import android.widget.Spinner;
import pl.karol202.paintplus.R;
import pl.karol202.paintplus.tool.ToolProperties;

public class SelectionProperties extends ToolProperties
{
	private ToolSelection selection;
	private SelectionShapeAdapter adapterShape;
	private SelectionModeAdapter adapterMode;
	
	private View view;
	private Spinner spinnerShape;
	private Spinner spinnerMode;
	private Button buttonSelectAll;
	private Button buttonSelectNothing;
	private Button buttonInvertSelection;
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
	{
		super.onCreateView(inflater, container, savedInstanceState);
		view = inflater.inflate(R.layout.properties_selection, container, false);
		setHasOptionsMenu(true);
		
		selection = (ToolSelection) tool;
		selection.setSelectionListener(new OnSelectionEditListener() {
			@Override
			public void onStartSelectionEditing()
			{
				getActivity().invalidateOptionsMenu();
			}
		});
		adapterShape = new SelectionShapeAdapter(getActivity());
		adapterMode = new SelectionModeAdapter(getActivity());
		
		spinnerShape = view.findViewById(R.id.spinner_selection_shape);
		spinnerShape.setAdapter(adapterShape);
		spinnerShape.setSelection(selection.getShape().ordinal());
		spinnerShape.setOnItemSelectedListener(new OnItemSelectedListener() {
			@Override
			public void onItemSelected(AdapterView<?> parent, View view, int position, long id)
			{
				onShapeSelected(position);
			}
			
			@Override
			public void onNothingSelected(AdapterView<?> parent)
			{
			
			}
		});
		
		spinnerMode = view.findViewById(R.id.spinner_selection_mode);
		spinnerMode.setAdapter(adapterMode);
		spinnerMode.setSelection(selection.getMode().ordinal());
		spinnerMode.setOnItemSelectedListener(new OnItemSelectedListener() {
			@Override
			public void onItemSelected(AdapterView<?> parent, View view, int position, long id)
			{
				onModeSelected(position);
			}
			
			@Override
			public void onNothingSelected(AdapterView<?> parent)
			{
			
			}
		});
		
		buttonSelectAll = view.findViewById(R.id.button_selection_all);
		buttonSelectAll.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v)
			{
				selection.selectAll();
			}
		});
		
		buttonSelectNothing = view.findViewById(R.id.button_selection_nothing);
		buttonSelectNothing.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v)
			{
				selection.selectNothing();
			}
		});
		
		buttonInvertSelection = view.findViewById(R.id.button_selection_invert);
		buttonInvertSelection.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v)
			{
				selection.invertSelection();
			}
		});
		
		return view;
	}
	
	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater)
	{
		if(selection.isInEditMode()) inflater.inflate(R.menu.menu_tool_selection, menu);
	}
	
	@Override
	public boolean onOptionsItemSelected(MenuItem item)
	{
		int id = item.getItemId();
		if(selection.isInEditMode())
		{
			switch(id)
			{
			case R.id.action_apply:
				selection.applySelection();
				break;
			case R.id.action_cancel:
				selection.cancelSelection();
				break;
			}
			getActivity().invalidateOptionsMenu();
			return true;
		}
		return super.onOptionsItemSelected(item);
	}
	
	private void onShapeSelected(int position)
	{
		ToolSelectionShape shape = ToolSelectionShape.values()[position];
		selection.setShape(shape);
	}
	
	private void onModeSelected(int position)
	{
		ToolSelectionMode mode = ToolSelectionMode.values()[position];
		selection.setMode(mode);
	}
}

File: app/src/main/java/pl/karol202/paintplus/tool/selection/SelectionProperties.kt
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package pl.karol202.paintplus.tool.selection

import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.Button
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import org.koin.android.ext.android.inject
import pl.karol202.paintplus.R
import pl.karol202.paintplus.databinding.PropertiesDragBinding
import pl.karol202.paintplus.databinding.PropertiesSelectionBinding
import pl.karol202.paintplus.image.ViewService
import pl.karol202.paintplus.options.OptionSelectAll
import pl.karol202.paintplus.options.OptionSelectInversion
import pl.karol202.paintplus.options.OptionSelectNothing
import pl.karol202.paintplus.tool.drag.ToolDrag
import pl.karol202.paintplus.util.collectIn
import pl.karol202.paintplus.util.setOnItemSelectedListener
import pl.karol202.paintplus.util.viewBinding

class SelectionProperties : Fragment(R.layout.properties_selection)
{
	private val toolSelection by inject<ToolSelection>()
	private val optionSelectAll by inject<OptionSelectAll>()
	private val optionSelectNothing by inject<OptionSelectNothing>()
	private val optionSelectInversion by inject<OptionSelectInversion>()

	private val views by viewBinding(PropertiesSelectionBinding::bind)

	override fun onViewCreated(view: View, savedInstanceState: Bundle?)
	{
		setHasOptionsMenu(true)
		toolSelection.updateEventFlow.collectIn(lifecycleScope) { activity?.invalidateOptionsMenu() }

		views.spinnerSelectionShape.adapter = SelectionShapeAdapter(requireContext())
		views.spinnerSelectionShape.setSelection(toolSelection.shape.ordinal)
		views.spinnerSelectionShape.setOnItemSelectedListener { onShapeSelected(it) }

		views.spinnerSelectionMode.adapter = SelectionModeAdapter(requireContext())
		views.spinnerSelectionMode.setSelection(toolSelection.mode.ordinal)
		views.spinnerSelectionMode.setOnItemSelectedListener { onModeSelected(it) }

		views.buttonSelectionAll.setOnClickListener { optionSelectAll.execute() }

		views.buttonSelectionNothing.setOnClickListener { optionSelectNothing.execute() }

		views.buttonSelectionInvert.setOnClickListener { optionSelectInversion.execute() }
	}

	override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater)
	{
		if(toolSelection.isInEditMode) inflater.inflate(R.menu.menu_tool_selection, menu)
	}

	override fun onOptionsItemSelected(item: MenuItem): Boolean
	{
		if(toolSelection.isInEditMode)
		{
			when(item.itemId)
			{
				R.id.action_apply -> toolSelection.applySelection()
				R.id.action_cancel -> toolSelection.cancelSelection()
			}
			return true
		}
		return super.onOptionsItemSelected(item)
	}

	private fun onShapeSelected(position: Int)
	{
		toolSelection.shape = ToolSelectionShape.values()[position]
	}

	private fun onModeSelected(position: Int)
	{
		toolSelection.mode = ToolSelectionMode.values()[position]
	}
}
